Add ReadOnlyMemory based overload for CanReadToken

henrikwidlund opened the pull request:
# Add Memory based overload for CanReadToken
<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the IdentityModel repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/dev/Contributing.md) and [Code of Conduct](https://opensource.microsoft.com/codeofconduct/).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [ ] If any gains or losses in performance are possible, you've included benchmarks for your changes. [More info](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki/Benchmarking-in-Wilson)
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->

## Description

Added an overload for the `CanReadToken` in `JsonWebTokenHandler` and `JwtSecurityTokenHandler` with `ReadOnlyMemory<char>` to reduce allocations and adjusted the `string` based overload to call the new method.

New overload is only available on dotnet 8 and above as the span based regex methods used to match JWS/JWE isn't available in dotnet 6.

Fixes #3221

henrikwidlund committed:
Add Span based overload for CanReadToken

Fixes #3221

henrikwidlund commented:
@microsoft-github-policy-service agree

henrikwidlund commented:
I did not add benchmarks, even though this is a performance improvement. The reason being that the string based API calls the span based one.

henrikwidlund committed:
Always include memory based overload to make shipped apis work

Change to ReadOnlyMemory to reduce allocations when calling ToString

kllysng commented:
Thanks for your contribution @henrikwidlund . Are you planning to add unit tests?

henrikwidlund commented:
@kllysng I was thinking that the regular string based ones would cover it since the string based method calls the memory based one. Let me know if you want explicit tests.

henrikwidlund committed:
Add to JwtSecurityTokenHandler + tests

henrikwidlund commented:
I've added tests now either way. Also added the method to `JwtSecurityTokenHandler`. Was thinking of adding the memory based overload to `JwtSecurityTokenHandler.ReadToken`, but feels out of scope.

brentschmaltz commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:
```diff
@@ -118,6 +118,59 @@ public IDictionary<string, string> InboundClaimTypeMap
+            // more segments than the maximum we know how to handle.
+            int segmentCount = JwtTokenUtilities.CountJwtTokenPart(token.Span, JwtConstants.MaxJwtSegmentCount + 1);
+
+            switch (segmentCount)
```
this code (switch (segmentCount)) can be shared between JsonWebTokenHandler and JwtSecurityTokenHandler.

When you think about it, most of this code could be shared between JsonWebTokenHandler and JwtSecurityTokenHandler.

henrikwidlund replied:
@brentschmaltz please resolve if you're satisfied with the solution :)

brentschmaltz reviewed (changes requested):
refactor and share common code between JsonWebTokenHandler and JwtSecurityTokenHandler.

henrikwidlund committed:
PR feedback

henrikwidlund commented:
Good points @brentschmaltz, I've applied your suggestions.

sruke committed:
Merge branch 'dev' into feature/jsonwebtokenhandler_memory

sruke reviewed (approved):
LGTM, thank you for taking the suggestions and making the contribution @henrikwidlund.

henrikwidlund commented:
I see that a few unit tests are failing because of accessibility. I'll make the method internal again tomorrow

henrikwidlund committed:
Make CountJwtTokenPart internal again as it's used in JwtSecurityTokenHandler

henrikwidlund committed:
Merge branch 'dev' into feature/jsonwebtokenhandler_memory

henrikwidlund commented:
@sruke fixed the compiler error

henrikwidlund committed:
Fix compiler issues and remove in keyword (might not work with dotnet framework.

sruke committed:
Merge branch 'dev' into feature/jsonwebtokenhandler_memory

sruke commented:
@henrikwidlund, a few unit tests are still failing. We’re planning to release IdentityModel today—would it be okay to include this PR in the next release (in a month) instead?

henrikwidlund committed:
Simplify logic and fix bug

henrikwidlund commented:
@sruke I believe I've fixed the test (was a real bug). I leave it up to you if it's getting to tight for your deadline (although, it would be nice to have it :))

henrikwidlund committed:
Merge branch 'dev' into feature/jsonwebtokenhandler_memory

henrikwidlund commented:
I noticed that the PR build is still failing, but I don't understand if it's related to my code changes?

sruke commented:
> I noticed that the PR build is still failing, but I don't understand if it's related to my code changes?

The failure (Resource not accessible by integration) appears unrelated to recent changes. All tests passed. I’m rerunning the build to check if the issue resolves. This change wasn’t included in the previous IdentityModel release but will be part of the upcoming one.

henrikwidlund commented:
> > I noticed that the PR build is still failing, but I don't understand if it's related to my code changes?
> 
> 
> 
> The failure (Resource not accessible by integration) appears unrelated to recent changes. All tests passed. I’m rerunning the build to check if the issue resolves. This change wasn’t included in the previous IdentityModel release but will be part of the upcoming one.

Notice that it didn't help 😕

henrikwidlund commented:
@sruke @brentschmaltz ping, any ideas what the problem is and when this can be merged?

henrikwidlund committed:
Merge remote-tracking branch 'origin/dev' into feature/jsonwebtokenhandler_memory

# Conflicts:
#	src/Microsoft.IdentityModel.JsonWebTokens/PublicAPI.Unshipped.txt

henrikwidlund commented:
@sruke @brentschmaltz ping again :/

henrikwidlund committed:
Merge branch 'dev' into feature/jsonwebtokenhandler_memory

brentschmaltz commented on src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.cs:
```diff
@@ -580,30 +580,72 @@ internal static SecurityKey ResolveTokenSigningKey(string kid, string x5t, IEnum
         /// <param name="maxCount">The maximum number of segments to count up to.</param>
         /// <returns>The number of segments up to <paramref name="maxCount"/>.</returns>
-        internal static int CountJwtTokenPart(string token, int maxCount)
+        internal static int CountJwtTokenPart(ReadOnlySpan<char> token, int maxCount)
```
Why do we need this as ReadOnlySpan?

We don't like to remove the old method as that can cause issues when mixed versions are used.
We end up with a MethodNotFound exception

henrikwidlund replied:
We need it because without it the PR would not make sense as we'd need to work with strings again. If you want I can create a method that takes string and calls this one.

brentschmaltz replied:
@henrikwidlund i was thinking why ReadOnlySpan and not ReadOnlyMemory?

henrikwidlund replied:
Only reasons are passing a smaller object and that that's what the method works with. If that's what you're thinking about I don't understand your comment about method not found as that would happen regardless.

torarnegustad replied:
@brentschmaltz I believe older versions of .NET does not support regular expressions for `ReadOnlySpan`, and this library targets older versions of .NET. This is the reason `ReadOnlyMemory` is used in the `CanReadToken` method.

`ReadOnlySpan` is more efficient and what libraries should use/support whenever possible.

I agreee that supporting `string` would be useful, even if it is basically just a wrapper for the new method.

brentschmaltz replied:
@henrikwidlund
> Only reasons are passing a smaller object and that that's what the method works with. If that's what you're thinking about I don't understand your comment about method not found as that would happen regardless.

If we remove a method in M.IM.Tokens in say, 8.81, then it is possible that M.IM.JsonWebTokens 8.80 might have a call that expected that method. This is particularly problematic for internal methods.

henrikwidlund replied:
So what's your suggestion, add back the string based one and have that one call the span one, or close the PR?

brentschmaltz replied:
> So what's your suggestion, add back the string based one and have that one call the span one, or close the PR?

Yes, that is what we are thinking.

henrikwidlund replied:
Which of them? I can do both, but it would be a bit of a waste 😅

brentschmaltz replied:
add back the string based one and have that one call the span one,

henrikwidlund replied:
Okay, done :)

henrikwidlund committed:
Add back string based version of CountJwtTokenPart for back compat

henrikwidlund committed:
Merge branch 'dev' into feature/jsonwebtokenhandler_memory

henrikwidlund commented:
Is this PR still wanted? It's been open for four months now. @brentschmaltz @sruke @jennyf19

pmaytak commented:
@henrikwidlund Sorry about the delay. I'm on holiday next week. I'll bring this up with the team, if there's available bandwidth; otherwise, I'll follow up on this after I'm back.
